src/invariant/test/OrderBookIsNotCrossedTests.cpp

```diff
@@ -0,0 +1,333 @@
+          "application behaviour",
+          "[invariant][OrderBookIsNotCrossed]")
+{
+    Asset cur1;
+    cur1.type(ASSET_TYPE_CREDIT_ALPHANUM4);
+    strToAssetCode(cur1.alphaNum4().assetCode, "CUR1");
+
+    Asset cur2;
+    cur2.type(ASSET_TYPE_CREDIT_ALPHANUM4);
+    strToAssetCode(cur2.alphaNum4().assetCode, "CUR2");
+
+    VirtualClock clock;
+    auto app = createTestApplication(clock, getTestConfig(0));
+    LedgerTxn ltxroot{app->getLedgerTxnRoot()};
+
+    auto const& invariant = std::make_shared<OrderBookIsNotCrossed>();
+    auto offer = generateOffer(cur1, cur2, 3, Price{3, 2});
+
+    // create
+    {
+        LedgerTxn ltx{ltxroot};
+        ltx.create(offer);
+
+        invariant->checkOnOperationApply({}, OperationResult{}, ltx.getDelta());
+        auto const& orders = invariant->getOrderBook().at(cur1).at(cur2);
+
+        REQUIRE(orders.size() == 1);
+
+        auto const& offerToCheck = *orders.cbegin();
+
+        REQUIRE(offerToCheck.amount == 3);
+        REQUIRE(offerToCheck.price.n == 3);
+        REQUIRE(offerToCheck.price.d == 2);
+
+        ltx.commit();
+    }
+
+    // modify
+    {
+        LedgerTxn ltx{ltxroot};
+
+        offer.data.offer().amount = 2;
+        offer.data.offer().price = Price{5, 3};
+        auto entry = ltx.load(LedgerEntryKey(offer));
+        entry.current() = offer;
+
+        invariant->checkOnOperationApply({}, OperationResult{}, ltx.getDelta());
+        auto const& orders = invariant->getOrderBook().at(cur1).at(cur2);
+
+        REQUIRE(orders.size() == 1);
+
+        auto const& offerToCheck = *orders.cbegin();
+
+        REQUIRE(offerToCheck.amount == 2);
+        REQUIRE(offerToCheck.price.n == 5);
+
+        ltx.commit();
+    }
+
+    // delete
+    {
+        LedgerTxn ltx{ltxroot};
+
+        auto entry = ltx.load(LedgerEntryKey(offer));
+        entry.erase();
+
+        invariant->checkOnOperationApply({}, OperationResult{}, ltx.getDelta());
+
+        REQUIRE(invariant->getOrderBook().at(cur1).at(cur2).size() == 0);
+    }
+}
+
+TEST_CASE("OrderBookIsNotCrossed properly throws if order book is crossed",
+          "[invariant][OrderBookIsNotCrossed]")
+{
+    Config cfg = getTestConfig(0);
+    cfg.INVARIANT_CHECKS = {"OrderBookIsNotCrossed"};
+
+    VirtualClock clock;
+    auto app = createTestApplication(clock, cfg);
+
+    Asset cur1;
+    cur1.type(ASSET_TYPE_CREDIT_ALPHANUM4);
+    strToAssetCode(cur1.alphaNum4().assetCode, "CUR1");
+
+    Asset cur2;
+    cur2.type(ASSET_TYPE_CREDIT_ALPHANUM4);
+    strToAssetCode(cur2.alphaNum4().assetCode, "CUR2");
+
+    Asset cur3;
+    cur3.type(ASSET_TYPE_CREDIT_ALPHANUM4);
+    strToAssetCode(cur3.alphaNum4().assetCode, "CUR3");
+
+    // the initial set up for the order book follows where:
+    // A = cur1, B = cur2, and C = cur3
+    // A against B
+    //     offer1: 3 A @  3/2 (B/A)
+    //     offer2: 2 A @  5/3 (B/A)
+    auto const& offer1 = createOffer(cur1, cur2, 3, Price{3, 2});
```

Comment:
reference again?

Reply:
this file is filled with bogus refs to temp variable, please fix

Reply:
Wow, those were weird.  I've gotten rid of all the ones I spotted.